File: algorithms/1250.longest_common_subsequence/longest_common_subsequence.cpp
class Solution {
public:
    int longestCommonSubsequence(const string &text1, const string &text2) {
        const int m = text1.size(), n = text2.size();
        vector<vector<int>> dp(m+1, vector(n+1, 0));
        for (int i = 1; i <= m; ++i) {
            const char &c1 = text1[i-1];
            for (int j = 1; j <= n; ++j) {
                const char &c2 = text2[j-1];
                if (c1 == c2) {
                    dp[i][j] = dp[i-1][j-1] + 1;
                } else {
                    dp[i][j] = max(dp[i][j-1], dp[i-1][j]);
                }
            }
        }
        return dp[m][n];
    }
};

File: algorithms/1250.longest_common_subsequence/longest_common_subsequence.rs
impl Solution {
    pub fn longest_common_subsequence(text1: String, text2: String) -> i32 {
        let (m, n) = (text1.len(), text2.len());
        let (text1, text2) = (text1.as_bytes(), text2.as_bytes());
        let (mut last, mut dp) = (vec![0; n + 1], vec![0; n + 1]);
        for i in 1..=m {
            dp.swap_with_slice(&mut last);
            for j in 1..=n {
                dp[j] = if text1[i - 1] == text2[j - 1] {
                    last[j - 1] + 1
                } else {
                    last[j].max(dp[j - 1])
                };
            }
        }
        dp[n]
    }
}